types/universal-router/index.d.ts

```diff
@@ -1,68 +1,118 @@
-// Type definitions for universal-router 1.2
+// Type definitions for universal-router 6.0
 // Project: https://github.com/kriasoft/universal-router
 // Definitions by: Jack Moore <https://github.com/jtmthf>
+//                 Tomek Łaziuk <https://github.com/tlaziuk>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.3
 
 /**
- * Params is a key/value object that represents extracted URL paramters. Each
- * URL parameter resolves to a string.
+ * Params is a key/value object that represents extracted URL paramters.
+ * Each URL parameter resolves to a string.
  */
 export interface Params {
-  [key: string]: string;
+    [_: string]: any;
 }
 
 /**
- * Context represents the context that is passed as the second argument
- * passed to resolve. By default, it only is require to contain a path, but
- * can be extended by the way of generics.
+ * In addition to a URL path string, any arbitrary data can be passed to the `router.resolve()` method, that becomes available inside action functions.
  */
 export interface Context {
-  path: string;
+    [_: string]: any;
 }
 
-/**
- * ActionContext is similar to Context, with the exception of an added params
- * object. ActionContext is passed as the first argument to the action
- * function.
- */
-export interface ActionContext extends Context {
-  params: Params;
+export interface PathnameContext extends Context {
+    /**
+     * URL which was transmitted to `router.resolve()`.
+     */
+    pathname: string;
+}
+
+export interface ActionContext<C extends Context, R = any> extends PathnameContext {
+    /**
+     * Current router instance.
+     */
+    router: UniversalRouter<C, R>;
+    /**
+     * Matched route object.
+     */
+    route: Route;
+    /**
+     * Middleware style function which can continue resolving.
+     */
+    next: (resume?: boolean, parent?: Route, prevResult?: any) => Promise<R>;
+    /**
+     * Base URL path relative to the path of the current route.
+     */
+    baseUrl: string;
+    /**
+     * Matched path.
+     */
+    path: string;
+    /**
+     * Matched path params.
+     */
+    params: Params;
+    /**
+     * An array of keys found in the path, see `path-to-regexp` documentation for details.
+     * @see https://github.com/pillarjs/path-to-regexp
+     */
+    keys: any;
```

Comment:
I would suggest to add `Key` type from `path-to-regexp` package. Since it has its own typescript declaration, it can be added as a separate dependency through `package.json`, as described [here](https://github.com/DefinitelyTyped/DefinitelyTyped#i-notice-some-packages-having-a-packagejson-here).